ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,530 @@
+    _Outptr_ void** context)
+{
+    EBPF_EXT_LOG_ENTRY();
+    ebpf_result_t result;
+    netevent_event_md_t* netevent_event_context = NULL;
+
+    *context = NULL;
+
+    if (context_in == NULL || context_size_in < sizeof(netevent_event_md_t)) {
+        EBPF_EXT_LOG_MESSAGE(EBPF_EXT_TRACELOG_LEVEL_ERROR, EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, "Context is required");
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    // Allocate memory for the context.
+    netevent_event_context = (netevent_event_md_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(netevent_event_md_t), EBPF_NETEVENT_EXTENSION_POOL_TAG);
+    EBPF_EXT_BAIL_ON_ALLOC_FAILURE_RESULT(
+        EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, netevent_event_context, "netevent_event_context", result);
+
+    // Copy the context from the caller.
+    memcpy(netevent_event_context, context_in, sizeof(netevent_event_md_t));
+
+    // Copy the event's pointer & size from the caller, to the out context.
+    netevent_event_context->event_data_start = (uint8_t*)data_in;
+    netevent_event_context->event_data_end = (uint8_t*)data_in + data_size_in;
+    *context = netevent_event_context;
+    netevent_event_context = NULL;
+    result = EBPF_SUCCESS;
+
+Exit:
+    if (netevent_event_context) {
+        ExFreePool(netevent_event_context);
+        netevent_event_context = NULL;
+    }
+    EBPF_EXT_RETURN_RESULT(result);
+}
+
+static void
+_ebpf_netevent_program_context_destroy(
+    _In_opt_ void* context,
+    _Out_writes_bytes_to_(*data_size_out, *data_size_out) uint8_t* data_out,
+    _Inout_ size_t* data_size_out,
+    _Out_writes_bytes_to_(*context_size_out, *context_size_out) uint8_t* context_out,
+    _Inout_ size_t* context_size_out)
+{
+    EBPF_EXT_LOG_ENTRY();
+
+    netevent_event_md_t* netevent_event_context = (netevent_event_md_t*)context;
+    netevent_event_md_t* netevent_event_context_out = (netevent_event_md_t*)context_out;
+
+    if (!netevent_event_context) {
+        goto Exit;
+    }
+
+    if (context_out != NULL && *context_size_out >= sizeof(netevent_event_md_t)) {
+        // Copy the context to the caller.
+        memcpy(netevent_event_context_out, netevent_event_context, sizeof(netevent_event_md_t));
+        *context_size_out = sizeof(netevent_event_md_t);
+
+        // Zero out the event context info.
+        netevent_event_context_out->event_data_start = 0;
+        netevent_event_context_out->event_data_end = 0;
+        *context_size_out = sizeof(netevent_event_md_t);
+    } else {
+        *context_size_out = 0;
+    }
+
+    // Copy the event data to 'data_out'.
+    if (data_out != NULL &&
+        *data_size_out >= (size_t)(netevent_event_context->event_data_end - netevent_event_context->event_data_start)) {
+        memcpy(
+            data_out,
+            netevent_event_context->event_data_start,
+            netevent_event_context->event_data_end - netevent_event_context->event_data_start + 1);
```

Comment:
Should this have the `+1`? I thought the event_data_end is the exclusive end of range.

Reply:
missed in the cleanup, thanks